RandBLAS/dense.hh

```diff
@@ -177,126 +173,103 @@ DenseSkOp<T>::~DenseSkOp() {
 
 
 
-
-
-
-
-template <typename T, typename T_gen>
-static RNGState gen_unif(
+/** Fill a n by m matrix with random values. If RandBLAS is compiled with
+ * OpenMP threading support enabled, the operation is parallelized using
+ * OMP_NUM_THREADS. The sequence of values genrated is not dependent on the
+ * number of OpenMP threads.
+ *
+ * @tparam T the data type of the matrix
+ * @tparam RNG a random123 CBRNG type
+ * @tparm OP an operator that transforms raw random values into matrix
+ *           elements. See RandBLAS::base::uneg11 and RandBLAS::base::boxmul.
+ *
+ * @param[in] n_rows the number of rows in the matrix
+ * @param[in] n_cols the number of columns in the matrix
+ * @param[in] mat a pointer to a contiguous region of memory with space for
+ *                n_rows*n_cols elements of type T. This memory will be filled
+ *                with random values.
+ * @param[in] seed A CBRNG state
+ *
+ * @returns the updated CBRNG state
+ */
+template <typename T, typename RNG, typename OP>
+auto fill_rmat(
     int64_t n_rows,
     int64_t n_cols,
     T* mat,
-    const RNGState &state
+    const RNGState<RNG> & seed
 ) {
+    RNG rng;
+    auto [c, k] = seed;
+
     int64_t dim = n_rows * n_cols;
-    int64_t i;
-    Random123_RNGState<T_gen> impl_state(state);
-    T_gen gen;
-    typedef typename T_gen::ctr_type ctr_type;
-    ctr_type rout;
-    for (i = 0; i + 3 < dim; i += 4) {
-        rout = gen(impl_state.ctr, impl_state.key);
-        mat[i] = r123::uneg11<T>(rout.v[0]);
-        mat[i + 1] = r123::uneg11<T>(rout.v[1]);
-        mat[i + 2] = r123::uneg11<T>(rout.v[2]);
-        mat[i + 3] = r123::uneg11<T>(rout.v[3]);
-        impl_state.ctr.incr(4);
-    }
-    rout = gen(impl_state.ctr, impl_state.key);
-    int32_t j = 0;
-    while (i < dim) {
-        mat[i] =  r123::uneg11<T>(rout.v[j]);
-        ++i;
-        ++j;
+    int64_t nit = dim / RNG::ctr_type::static_size;
+    int64_t nlast = dim % RNG::ctr_type::static_size;
+
+#if defined(RandBLAS_HAS_OpenMP)
+    #pragma omp parallel firstprivate(c, k)
+    {
+        // add the start index to the counter in order to make the sequence
+        // deterministic independent of the number of threads.
+        int ti = omp_get_thread_num();
+        int nt = omp_get_num_threads();
+
+        int64_t chs = nit / nt;
+        int64_t nlg = nit % nt;
+        int64_t i0 = chs * ti + (ti < nlg ? ti : nlg);
+        int64_t i1 = i0 + chs + (ti < nlg ? 1 : 0);
```

Comment:
Can you give the full names behind these abbreviated variable names, via comments? Like:
```
        // chs = [word that starts with c] ... [other words] ...
        int64_t chs = nit / not;
        // nlg = [word that starts with l] ... [other words] ...
        int64_t nlg = nit % nt;
```

Reply:
lol, sorry about that. I'll fix this, here chs means chunk_size, it's a term used in the OMP spec

Reply:
nlg is number of large chunks

Reply:
resolved in 02977485